hil/ext/switches/juniper.py

```diff
@@ -0,0 +1,434 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS
+# IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied.  See the License for the specific language
+# governing permissions and limitations under the License.
+"""A switch driver for the juniper QFX series.
+
+But in theory can work for any Juniper switch that supports NETCONF.
+Driver uses the python library PyEZ made available by Juniper.
+By default it sets NETCONF session over ssh, i.e. make RPC calls over ssh.
+For this to work, make sure that the Juniper switch is configured
+appropriately to recieve NETCONF connections. Its default port is 830.
+"""
+
+import re
+import logging
+import schema
+import json
+import sys
+import os
+
+from jnpr.junos import Device
+from jnpr.junos.exception import ConnectError
+from jnpr.junos.utils.config import Config
+from jnpr.junos.utils.config import ConfigLoadError
+
+
+from hil.model import db, Switch, BigIntegerType
+from hil.migrations import paths
+from hil import config
+from hil.config import cfg
+from os.path import dirname, join
+
+import pdb; pdb.set_trace()
+config.load()
+j_driver = 'hil.ext.switches.juniper'
+if cfg.has_option('extensions', j_driver):
+    template_dir = cfg.get('extensions', j_driver)
+    sys.path.append(template_dir)
+    try:
+        from config_tables.ConfigTables import InterfaceConfigTable
+    except ImportError:
+        pass
+
+logger = logging.getLogger(__name__)
+paths[__name__] = join(dirname(__file__), 'migrations', 'juniper')
+
+
+class VlanAddError(Exception):
+    """ Raise this exception when vlan cannot be added to the port."""
+    pass
+
+
+class ConfigCommitError(Exception):
+    """ Raise this exception if there are commit conflicts."""
+    pass
+
+
+class Juniper(Switch):
+    api_name = 'http://schema.massopencloud.org/haas/v0/switches/juniper'
+    dir_name = os.path.dirname(__file__)
+    __mapper_args__ = {
+        'polymorphic_identity': api_name,
+    }
+
+    id = db.Column(
+            BigIntegerType, db.ForeignKey('switch.id'), primary_key=True
+            )
+    hostname = db.Column(db.String, nullable=False)
+    username = db.Column(db.String, nullable=False)
+    password = db.Column(db.String, nullable=False)
+
+    enable_interface = None
+    add_vlan = None
+    remove_vlan = None
+    set_native_vlan = None
+    remove_native_vlan = None
+    remove_default_vlan = None
+    set_trunk_mode = None
+
+    @staticmethod
+    def validate(kwargs):
+        schema.Schema({
+            'username': basestring,
+            'hostname': basestring,
+            'password': basestring,
+        }).validate(kwargs)
+
+    def session(self):
+        return self
+
+    def disconnect(self):
+        pass
+
+    @staticmethod
+    def validate_port_name(port):
+        pass
+
+    def _create_session(self):
```

Comment:
Looks like the switch is connection oriented. In that case your driver should look like how it is defined in `hil/model.py` (or like the nexus driver). This is important because the Networking Daemon caches the session object which avoids connecting-disconnecting for each change.